docs/rfc/0001_ctrlkit.md

```diff
@@ -0,0 +1,325 @@
+            owned
+        }
+    }
+
+    action {
+        // List all active jobs, and update the status.
+        ListActiveJobsAndUpdateStatus(jobs)
+
+        // Clean up old jobs according to the history limit.
+        CleanUpOldJobsExceedsHistoryLimits(jobs)
+
+        // Run the next job if it's on time, or otherwise we should wait 
+        // until the next scheduled time.
+        RunNextScheduledJob()
+
+        // Update status of CronJob.
+        UpdateCronJobStatus()
+    }
+}
+```
+
+### **CLI Tool**
+
+The CLI tool takes a document file written in DSL above and generates the polyfill codes. Currently, only Go is going to be supported as the target language.
+
+It generates the state getters and action wrappers by encapsulating them into structs and methods. It leaves the implementation of actions to the user with a stub interface. For example, it generates an output like this for the `CronJob` document above:
+
+```go
+type CronJobControllerManagerState struct {
+	client.Reader
+	target *apiv1.CronJob
+}
+
+func (s *CronJobControllerManagerState) GetJobs(ctx context.Context) ([]batchv1.Job, error) {
+	...
+}
+
+type CronJobControllerManagerImpl interface {
+	ctrlkit.CrontollerManagerActionLifeCycleHook
+
+	ListActiveJobsAndUpdateStatus(ctx context.Context, logger logr.Logger, jobs []batchv1.Job) (ctrl.Result, error)
+
+	CleanUpOldJobsExceedsHistoryLimits(ctx context.Context, logger logr.Logger, jobs []batchv1.Job) (ctrl.Result, error)
+
+	RunNextScheduledJob(ctx context.Context, logger logr.Logger) (ctrl.Result, error)
+
+	UpdateCronJobStatus(ctx context.Context, logger logr.Logger) (ctrl.Result, error)
+}
+
+type CronJobControllerManager struct {
+	state  CronJobControllerManagerState
+	impl   CronJobControllerManagerImpl
+	logger logr.Logger
+}
+
+func (m *CronJobControllerManager) ListActiveJobsAndUpdateStatus() ctrlkit.ReconcileAction {
+	return ctrlkit.WrapAction("ListActiveJobsAndUpdateStatus", func(ctx context.Context) (ctrl.Result, error) {
+		logger := m.logger.WithValues("action", "ListActiveJobsAndUpdateStatus")
+
+		jobs, err := m.state.GetJobs(ctx)
+		if err != nil {
+			return ctrlkit.RequeueIfError(err)
+		}
+
+		defer m.impl.AfterActionRun("ListActiveJobsAndUpdateStatus", ctx, logger)
+		m.impl.BeforeActionRun("ListActiveJobsAndUpdateStatus", ctx, logger)
+
+		return m.impl.ListActiveJobsAndUpdateStatus(ctx, logger, jobs)
+	})
+}
+
+...
+```
+
+The developer is responsible for writing an implementation of the `CronJobControllerManagerImpl` to implement the actions.
+
+### **Library**
+
+The library should contain three main parts:
+* Abstraction of actions `ReconcileAction`.
+* Functions for organizing the workflow `Join`, `Sequential`, `Timeout`, and `If`.
+* Functions and errors for continuing or interrupting the workflow.
+
+```go
+type ReconcileAction interface {
+    Description() string
+	Run(ctx context.Context) (ctrl.Result, error)
+}
+```
+
+**Workflow organizing functions**
```

Comment:
How to specify the workflow dependencies using DSL?

Reply:
Currently, there's no way. We organize the workflow with functions in the library.